engine/CSVParser.cc

```diff
@@ -447,26 +461,85 @@ namespace minsky
+		
+		// for feature 47
+		if (v.sparsityRatio <= 0.5) 
+	    {
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numDenseElements()*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");
+		  v.index.clear();	            
+          // stash the data into vv tensorInit field
+          v.tensorInit.data.clear();
+          v.tensorInit.data.resize(v.numDenseElements(), spec.missingValue);  
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+              v.tensorInit.data[idx]=i.second;  
+            }
+        }    
+        else 
+	    {	
+		  v.numSparseElements=tmpData.size();	
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numSparseElements*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");		  
+		  v.index.clear();			
+		  v.tensorInit.data.clear();	
+		  v.tensorInit.data.resize(v.numSparseElements, spec.missingValue); 	
+		  vector<double> tempData;
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+               if (!isnan(i.second)) {
+                 v.index.push_back(idx);
+                 tempData.push_back(i.second);
+               }                   
+            }
+           
+            // for feature 47           
+            for (auto& i: tmpAll)
+            {
+			  size_t hIdx=0;	
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  hIdx = (hIdx*dims[j]) + dimLabels[j][i.first[j]];
+                }                   
+                auto idx = find(v.index.begin(),v.index.end(),hIdx);   
+                // For feature 47. Stores only non-Nan elements from data in tensorInit.data                 
+                if (idx != v.index.end() && *idx<v.tensorInit.data.size()) v.tensorInit.data[*idx]=tempData[*idx];  
+             }               
+            
+            tempData.clear();   
+            
+            assert(tempData.size()*v.tensorInit.data.size()==0);
```

Comment:
This is guaranteed by the previous statement.

Reply:
Thanks, I've removed it.